### PR TITLE
Fix int types

### LINKED ISSUE #1913: Use correct integer types in Cython
Currently incorrect integer types are used in Cython implementation.
It's less portable and perhaps more importantly, confusing for developers.

For example:
* Pointers should be `intptr_t`.
* Memory sizes should be `size_t`.
* Memory offsets should be `ptrdiff_t`.

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #1913

The added benefit is that Cython now checks negative size for malloc:
```
>>> a = cupy.cuda.memory.alloc(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cupy/cuda/memory.pyx", line 506, in cupy.cuda.memory.alloc
    cpdef MemoryPointer alloc(size):
  File "cupy/cuda/memory.pyx", line 518, in cupy.cuda.memory.alloc
    return _current_allocator(size)
  File "cupy/cuda/memory.pyx", line 1068, in cupy.cuda.memory.MemoryPool.malloc
    cpdef MemoryPointer malloc(self, size_t size):
OverflowError: can't convert negative value to size_t
```

### CODE REVIEW

File: cupy/cuda/memory.pyx [hunk 1]
```diff
@@ -196,8 +196,8 @@ cdef class _Chunk:
```
Comment (asi1024) [inline cupy/cuda/memory.pyx]: Could you also fix docstring?

Comment (niboshi) [reply cupy/cuda/memory.pyx]: Fixed.

File: cupy/cuda/memory.pyx [hunk 2]
```diff
@@ -207,18 +207,18 @@ cdef class _Chunk:
         mem, offset, size, stream_ptr = args
         self._init(mem, offset, size, stream_ptr)
 
-    cdef _init(self, BaseMemory mem, Py_ssize_t offset,
-               Py_ssize_t size, Py_ssize_t stream_ptr):
-        assert mem.ptr > 0 or offset == 0
+    cdef _init(self, BaseMemory mem, ptrdiff_t offset,
+               size_t size, Py_ssize_t stream_ptr):
```
[the remark this answers is not in the document]
Comment (niboshi) [reply cupy/cuda/memory.pyx]: No. I want to fix it, too, but I want it to be another PR, because it should be `ctypedef`ed as some type like `stream_t`.

**Comment by asi1024** [comment]
jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**